corellium/shard/src/main/kotlin/flank/corellium/shard/Internal.kt

```diff
@@ -0,0 +1,105 @@
+package flank.corellium.shard
+
+import kotlin.math.min
+
+// Stage 1
+
+/**
+ * Create flat and easy to iterate list of [Chunk] without loosing info about app and test related to test case.
+ * @receiver The [List] of [Shard.App].
+ * @return The [List] of [Chunk] which is just different representation of input data.
+ */
+internal fun List<Shard.App>.mapToInternalChunks(): List<Chunk> =
+    mutableListOf<Chunk>().also { result ->
+        forEach { app ->
+            app.tests.forEach { test ->
+                test.cases.forEach { case ->
+                    result.add(
+                        Chunk(
+                            app = app.name,
+                            test = test.name,
+                            case = case.name,
+                            duration = case.duration
+                        )
+                    )
+                }
+            }
+        }
+    }
+
+/**
+ * Internal intermediate structure which is representing test case with associated app and test module.
+ */
+internal class Chunk(
+    val app: String,
+    val test: String,
+    val case: String,
+    val duration: Long,
+)
+
+// Stage 2
+
+/**
+ * Group the chunks into sub-lists where the standard deviation of group duration should by possibly small.
+ * @receiver The flat [List] of [Chunk].
+ * @return The [List] of [Chunk] groups balanced by the summary duration of each.
+ */
+internal fun List<Chunk>.groupByDuration(maxCount: Int): List<List<Chunk>> {
+    class Chunks(
+        var duration: Long = 0,
+        val list: MutableList<Chunk> = mutableListOf()
+    )
+
+    // The real part of sharding calculations,
+    // everything else is just a mapping between structures.
+    // ===================================================
+    return sortedByDescending(Chunk::duration).fold(
+        initial = List(min(size, maxCount)) { Chunks() }
+    ) { acc, chunk ->
+        acc.first().apply {
+            duration += chunk.duration
+            list += chunk
+        }
+        acc.sortedBy(Chunks::duration)
+    }.map(Chunks::list)
+    // ===================================================
+}
+
+// Stage 3
+
+/**
+ * Build the final structure of shards.
+ */
+internal fun List<List<Chunk>>.mapToShards(): List<List<Shard.App>> {
+
+    // Structure the group of chunks mapping them by app and test.
+    val list: List<Map<String, Map<String, List<Chunk>>>> = map { group ->
+        group.groupBy { chunk ->
+            chunk.app
+        }.mapValues { (_, chunks) ->
+            chunks.groupBy { chunk ->
+                chunk.test
+            }
+        }
+    }
+
+    // Convert grouped chunks into the output structures.
+    return list.map { map ->
+        map.map { (app, tests) ->
```

Comment:
I know it might be necessary but 4 nested maps is crazy...

Reply:
Nope, it's not crazy, it's just mapping one data representation to another without any condition. 
The crazy will be adding the conditions inside these mappings or creating improper data structures just because of some additional mappings and loops in private or internal functions.